ceviche/fdfd.py

```diff
@@ -191,6 +191,123 @@ def _solve_fn(self, eps_vec, entries_a, indices_a, Jz_vec):
+
+        C = sp.kron(sp.eye(M), - 1 / MU_0 * self.Dxf.dot(self.Dxb) - 1 / MU_0 * self.Dyf.dot(self.Dyb))
+        entries_c, indices_c = get_entries_indices(C)
+
+        # indices into the diagonal of a sparse matrix 
+        D = - EPSILON_0 * sp.kron(sp.spdiags(W**2,[0],M,M), sp.spdiags(eps_vec,[0],N,N))
+        entries_diag, indices_diag = get_entries_indices(D)
+
+        entries_a = npa.hstack((entries_diag, entries_c))
+        indices_a = npa.hstack((indices_diag, indices_c))
+
+        Nfreq = npa.shape(delta_matrix)[0]
+        for k in npa.arange(Nfreq):
+            delta_diag_p = - EPSILON_0 * sp.spdiags(delta_matrix[k,:] * npa.exp(1j*phi_matrix[k,:]), [0], N, N)
+            delta_diag_m = - EPSILON_0 * sp.spdiags(delta_matrix[k,:] * npa.exp(-1j*phi_matrix[k,:]), [0], N, N)
+            Z = sp.kron(sp.spdiags(0.5*npa.concatenate([ [0], W[:-(k+1)] ])**2,[k+1],M,M), delta_diag_p)
+            entries_z, indices_z = get_entries_indices(Z)
+            entries_a = npa.hstack((entries_z, entries_a))
+            indices_a = npa.hstack((indices_z,indices_a))
+            Z = sp.kron(sp.spdiags(0.5*npa.concatenate([ W[k+1:], [0] ])**2,[-(k+1)],M,M), delta_diag_m)
+            entries_z, indices_z = get_entries_indices(Z)
+            entries_a = npa.hstack((entries_z, entries_a))
+            indices_a = npa.hstack((indices_z,indices_a))
+
+        return entries_a, indices_a
+
+    def _solve_fn(self, eps_vec, entries_a, indices_a, Jz_vec):
+        M = 2*self.Nsb + 1
+        N = self.Nx * self.Ny
+        W = self.omega + npa.arange(-self.Nsb,self.Nsb+1)*self.Omega 
+        P = sp.kron(sp.spdiags(W,[0],M,M), sp.eye(N))
+        entries_p, indices_p = get_entries_indices(P)
+        b_vec = 1j * sp_mult(entries_p,indices_p,Jz_vec)
+        Ez_vec = sp_solve(entries_a, indices_a, b_vec)
+        Hx_vec, Hy_vec = self._Ez_to_Hx_Hy(Ez_vec)
+        return Hx_vec, Hy_vec, Ez_vec
+
+    def _Ez_to_Hx(self, Ez_vec):
+        M = 2*self.Nsb + 1
+        Winv = npa.reciprocal(self.omega + npa.arange(-self.Nsb,self.Nsb+1)*self.Omega)
+        Dyb_mf = sp.kron(sp.spdiags(Winv,[0],M,M), self.Dyb)
+        entries_Dyb_mf, indices_Dyb_mf = get_entries_indices(Dyb_mf)
+        return -1 / 1j / MU_0 * sp_mult(entries_Dyb_mf, indices_Dyb_mf, Ez_vec)
+
+    def _Ez_to_Hy(self, Ez_vec):
+        M = 2*self.Nsb + 1
+        Winv = npa.reciprocal(self.omega + npa.arange(-self.Nsb,self.Nsb+1)*self.Omega)
+        Dxb_mf = sp.kron(sp.spdiags(Winv,[0],M,M), self.Dxb)
+        entries_Dxb_mf, indices_Dxb_mf = get_entries_indices(Dxb_mf)
+        return  1 / 1j / MU_0 * sp_mult(entries_Dxb_mf, indices_Dxb_mf, Ez_vec)
+
+    def _Ez_to_Hx_Hy(self, Ez_vec):
+        Hx_vec = self._Ez_to_Hx(Ez_vec)
+        Hy_vec = self._Ez_to_Hy(Ez_vec)
+        return Hx_vec, Hy_vec
+
+    def _vec_to_grid(self, vec):
+        grid_shape = (2*self.Nsb + 1, self.Nx, self.Ny)
+        return npa.reshape(vec, grid_shape)
+
```

Comment:
For the notebook (dont know how to comment on it from here).  It looks great!  I'm wondering if you can just add a few markdown cells or some other comments explaining what's going on?   Also, you'll want to change the sys.path eventually because its specific to your computer now.

Reply:
Done!